project-code/core/common/src/main/scala/play/core/server/servlet/RequestHandler.scala

```diff
@@ -116,124 +127,140 @@ trait HttpServletRequestHandler extends RequestHandler {
-                      }
-                    })
-                    .map { _ => if (!hasError.get) Cont(step) else Done((), Input.Empty: Input[Array[Byte]]) }
-                    .andThen {
-                      case Failure(ex) =>
-                        hasError.set(true)
-                        Logger("play").debug(ex.toString)
-                        throw ex
-                    })
-              case (true, Input.Empty) => Cont(step)
-              case (_, inp) => Done((), inp)
-            }
-            Iteratee.flatten(
-              Future.successful(())
-                .map(_ => if (!hasError.get) Cont(step) else Done((), Input.Empty: Input[Array[Byte]])))
-          }
-
-          val bodyConsumer = if (chunked) {
-            // if the result body is chunked, the chunks are already encoded with metadata in Results.chunk
-            // The problem is that the servlet container adds metadata again, leading the chunks encoded 2 times.
-            // As workaround, we 'dechunk' the body one time before sending it to the servlet container
-            body &> Results.dechunk |>>> bodyIteratee
-          } else {
-            body |>>> bodyIteratee
-          }
-          bodyConsumer.andThen {
-            case Success(_) =>
-              cleanup()
-              onHttpResponseComplete()
-            case Failure(ex) =>
-              Logger("play").debug(ex.toString)
-              hasError.set(true)
-              onHttpResponseComplete()
-          }
-        } else {
-          Logger("play").trace("Result without Content-length")
-
-          // No Content-Length header specified, buffer in-memory
-          val byteBuffer = new ByteArrayOutputStream
-          val byteArrayOSIteratee = Iteratee.fold(byteBuffer)((b, e: Array[Byte]) => {
-            b.write(e); b
-          })
-
-          val p = body |>>> Enumeratee.grouped(byteArrayOSIteratee) &>> Cont {
-            case Input.El(buffer) =>
-              Logger("play").trace("Buffer size to send: " + buffer.size)
-              getHttpResponse().getHttpServletResponse.map { response =>
-                // set the content length ourselves
-                response.setContentLength(buffer.size)
-                val os = response.getOutputStream
-                os.flush()
-                buffer.writeTo(os)
-              }
-              val p = Future.successful(())
-              Iteratee.flatten(p.map(_ => Done(1, Input.Empty: Input[ByteArrayOutputStream])))
-
-            case other => Error("unexpected input", other)
-          }
-          p.andThen {
-            case Success(_) =>
-              cleanup()
-              onHttpResponseComplete()
-            case Failure(ex) =>
-              Logger("play").debug(ex.toString)
-              onHttpResponseComplete()
+        val flow = body.dataStream.map { s =>
+          getHttpResponse().getRichOutputStream.foreach { os =>
+            os.write(s.toArray)
+            os.flush()
```

Comment:
Same as previous comment, you're using blocking IO here, blocking on the application thread which will lead to thread exhaustion, this is a regression. Should use the ResponseSubscriber to convert the response to a Sink, and then you can feed dataStream into the response body.

Reply:
I was unable to use reactive-streams-servlet... :( 
`Sink.fromSubscriber(new ResponseSubscriber(asyncContext))` produces `Sink[ByteBuffer, NotUsed]` and I need (or at least I think I need) `Sink[ByteBuffer, Future[Any]]`.
Do you have any suggestion what to do with that?